File: store/internal/infra/middlewares/jwt_auth.go
package middlewares

import (
	"context"
	"crypto/rsa"
	"encoding/base64"
	"fmt"
	"log/slog"
	"math/big"
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/golang-jwt/jwt"

	auth "github.com/Jibaru/ichibuy/api-client/go/auth"

	sharedCtx "ichibuy/store/internal/shared/context"
)

type JWTAuthMiddleware struct {
	authClient *auth.APIClient
	cache      map[string]*rsa.PublicKey
}

func NewJWTAuthMiddleware(authClient *auth.APIClient) *JWTAuthMiddleware {
	return &JWTAuthMiddleware{
		authClient: authClient,
		cache:      make(map[string]*rsa.PublicKey),
	}
}

func (m *JWTAuthMiddleware) ValidateToken() gin.HandlerFunc {
	return func(c *gin.Context) {
		authHeader := c.GetHeader("Authorization")
		if authHeader == "" {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "Authorization header is required"})
			c.Abort()
			return
		}

		parts := strings.Split(authHeader, " ")
		if len(parts) != 2 || parts[0] != "Bearer" {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid authorization header format"})
			c.Abort()
			return
		}

		tokenString := parts[1]

		token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
			kid, ok := token.Header["kid"].(string)
			if !ok {
				return nil, fmt.Errorf("kid not found in token header")
			}

			publicKey, err := m.getPublicKey(c, kid)
			if err != nil {
				return nil, err
			}

			return publicKey, nil
		})

		if err != nil {
			c.JSON(http.StatusUnauthorized, gin.H{"error": fmt.Sprintf("Invalid token: %v", err)})
			c.Abort()
			return
		}

		if !token.Valid {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "Token is not valid"})
			c.Abort()
			return
		}

		claims, ok := token.Claims.(jwt.MapClaims)
		if !ok {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid token claims"})
			c.Abort()
			return
		}

		userID, ok := claims["user_id"].(string)
		if !ok || userID == "" {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "user_id not found in token"})
			c.Abort()
			return
		}

		slog.InfoContext(c, "token validated successfully", "user_id", userID)

		c.Set("user_id", userID)
		c.Set(sharedCtx.APITokenKey, tokenString)
		c.Next()
	}
}

func (m *JWTAuthMiddleware) getPublicKey(c context.Context, kid string) (*rsa.PublicKey, error) {
	if cachedKey, exists := m.cache[kid]; exists {
		return cachedKey, nil
	}

	jwks, _, err := m.authClient.DefaultApi.ApiV1AuthWellKnownJwksJsonGet(c)
	if err != nil {
		return nil, err
	}

	for _, key := range jwks.Keys {
		if key.Kid == kid {
			publicKey, err := m.jwkToPublicKey(key)
			if err != nil {
				return nil, err
			}

			m.cache[kid] = publicKey
			return publicKey, nil
		}
	}

	return nil, fmt.Errorf("public key not found for kid: %s", kid)
}

func (m *JWTAuthMiddleware) jwkToPublicKey(jwk auth.HandlersJwk) (*rsa.PublicKey, error) {
	nBytes, err := base64.URLEncoding.DecodeString(jwk.N)
	if err != nil {
		return nil, fmt.Errorf("failed to decode n: %w", err)
	}

	eBytes, err := base64.URLEncoding.DecodeString(jwk.E)
	if err != nil {
		return nil, fmt.Errorf("failed to decode e: %w", err)
	}

	n := new(big.Int).SetBytes(nBytes)
	e := new(big.Int).SetBytes(eBytes)

	publicKey := &rsa.PublicKey{
		N: n,
		E: int(e.Int64()),
	}

	return publicKey, nil
}

File: fstorage/src/middlewares/jwt_auth.ts
import { Request, Response, NextFunction } from 'express';
import jwt from 'jsonwebtoken';
import * as crypto from 'crypto';

interface JWK {
  kid: string;
  n: string;
  e: string;
  kty: string;
  use: string;
}

interface JWKS {
  keys: JWK[];
}

interface AuthenticatedRequest extends Request {
  userId?: string;
}

class JWTAuthMiddleware {
  private authBaseUrl: string;
  private cache: Map<string, string>;

  constructor(authBaseUrl: string) {
    this.authBaseUrl = authBaseUrl;
    this.cache = new Map<string, string>();
  }

  public validateToken() {
    return async (req: AuthenticatedRequest, res: Response, next: NextFunction) => {
      try {
        const authHeader = req.headers.authorization;
        if (!authHeader) {
          return res.status(401).json({ error: 'Authorization header is required' });
        }

        const parts = authHeader.split(' ');
        if (parts.length !== 2 || parts[0] !== 'Bearer') {
          return res.status(401).json({ error: 'Invalid authorization header format' });
        }

        const tokenString = parts[1];

        const payload = await new Promise<any>((resolve, reject) => {
          jwt.verify(tokenString, async (header, callback) => {
            try {
              const kid = header.kid;
              if (!kid) {
                return callback(new Error('kid not found in token header'));
              }

              const publicKey = await this.getPublicKey(kid);
              if (!publicKey) {
                return callback(new Error(`Public key not found for kid: ${kid}`));
              }

              callback(null, publicKey);
            } catch (error) {
              callback(error as Error);
            }
          }, { algorithms: ['RS256'] }, (error, decoded) => {
            if (error) {
              reject(error);
            } else {
              resolve(decoded);
            }
          });
        });

        if (!payload || !payload.user_id) {
          return res.status(401).json({ error: 'user_id not found in token' });
        }

        console.log('Token validated successfully', { user_id: payload.user_id });
        req.userId = payload.user_id as string;
        next();
      } catch (error) {
        console.error('JWT validation error:', error);
        return res.status(401).json({ error: `Invalid token: ${error instanceof Error ? error.message : 'Unknown error'}` });
      }
    };
  }

  private async getPublicKey(kid: string): Promise<string | null> {
    if (this.cache.has(kid)) {
      return this.cache.get(kid)!;
    }

    try {
      const jwksUrl = `${this.authBaseUrl}/api/v1/auth/.well-known/jwks.json`;
      const response = await fetch(jwksUrl);

      if (!response.ok) {
        throw new Error(`Failed to fetch JWKS: ${response.status}`);
      }

      const jwks: JWKS = await response.json();

      for (const key of jwks.keys) {
        if (key.kid === kid) {
          const publicKey = this.jwkToPem(key);
          this.cache.set(kid, publicKey);
          return publicKey;
        }
      }

      throw new Error(`Public key not found for kid: ${kid}`);
    } catch (error) {
      console.error('Error fetching public key:', error);
      return null;
    }
  }

  private jwkToPem(jwk: JWK): string {
    try {
      const keyObject = crypto.createPublicKey({
        key: {
          kty: 'RSA',
          n: jwk.n,
          e: jwk.e
        },
        format: 'jwk'
      });

      return keyObject.export({ type: 'spki', format: 'pem' }) as string;
    } catch (error) {
      console.error('Error converting JWK to PEM:', error);
      throw error;
    }
  }
}

export { JWTAuthMiddleware, AuthenticatedRequest };